ports/rp2/machine_pwm.c

```diff
@@ -165,14 +244,24 @@ STATIC void mp_machine_pwm_freq_set(machine_pwm_obj_t *self, mp_int_t freq) {
     }
     pwm_hw->slice[self->slice].div = div16;
     pwm_hw->slice[self->slice].top = top;
+    slice_freq_set[self->slice] = true;
     if (self->duty_type == DUTY_U16) {
         mp_machine_pwm_duty_set_u16(self, self->duty);
     } else if (self->duty_type == DUTY_NS) {
         mp_machine_pwm_duty_set_ns(self, self->duty);
     }
+    machine_pwm_obj_t *other = self->channel == PWM_CHAN_A ? self + 1 : self - 1;
+    if (other->duty_type == DUTY_U16) {
+        mp_machine_pwm_duty_set_u16(other, other->duty);
+    } else if (other->duty_type == DUTY_NS) {
+        mp_machine_pwm_duty_set_ns(other, other->duty);
+    }
 }
 
 STATIC mp_obj_t mp_machine_pwm_duty_get_u16(machine_pwm_obj_t *self) {
+    if (self->duty_type == DUTY_NOT_SET || slice_freq_set[self->slice] == false) {
+        mp_raise_ValueError(MP_ERROR_TEXT("PWM not configured"));
+    }
```

Comment:
I don't like this situation where then PWM module has been initialized and is doing *something*, but I'm not allowed to ask what it's doing:

```python
from machine import PWM, Pin
p=PWM(Pin(25)) # Pin 25 goes low, LED turns on
p.duty_u16(65535) # duty cycle is still 0%
p.duty_u16() # error
p.freq() # error
```

Just as an example of how this could be a disaster in real applications, imagine a display update that happens somehow asynchronously and checks `duty_u16()` as part of the update cycle for some reason. Maybe it's a servo controller that outputs the current servo settings periodically to a serial port. Usually it's fine but 1 out of 1000 times the display update happens before duty/freq are both set. Now we'll have an error maybe causing a crash in the field where it can't be debugged.

Note that this hypothetical application might already be written. We don't have a way to notify the author of the new requirements, and this update will silently cause their code (which worked fine before) to become unreliable.

It seems like it would be much better to allow checking PWM state without errors.  I would much prefer one of these three options:

* start at a reasonable default frequency (the current behavior for rp2)
* return zero duty cycle and frequency when PWM is off
* require setting an initial frequency and duty cycle in the constructor

As a comparison, Arduino's [analogWrite()](https://www.arduino.cc/reference/en/language/functions/analog-io/analogwrite/) function uses a preconfigured frequency. I had expected MicroPython to work the same way when I got the surprising result I mentioned in #11336.

Reply:
The duty and freq methods no longer raise exceptions when queried, but instead return 0 if not started yet.